transaction/std/src/memo/gift_code_cancellation.rs

```diff
@@ -23,17 +25,53 @@ impl RegisteredMemoType for GiftCodeCancellationMemo {
 }
 
 impl GiftCodeCancellationMemo {
-    /// The length of the custom memo data
+    /// Number of bytes in the memo
     pub const MEMO_DATA_LEN: usize = 64;
 
-    /// Create a new gift code memo
-    pub fn new(global_index: u64) -> Self {
-        GiftCodeCancellationMemo::from(global_index)
+    /// Number of bytes used to represent the global index of the TxOut used to
+    /// fund the cancelled gift code
+    pub const INDEX_DATA_LEN: usize = 8;
+
+    /// Number of bytes used to represent the fee paid for gift code
+    /// cancellation
+    pub const FEE_DATA_LEN: usize = 7;
+
+    /// Maximum value of the fee we can record
+    pub const MAX_FEE: u64 = u64::MAX >> 8;
+
+    /// Create a new gift code cancellation memo
+    pub fn new(global_index: u64, fee: u64) -> Result<Self, MemoError> {
+        // Check if the fee we're setting is greater than the max fee
+        if fee > Self::MAX_FEE {
+            return Err(MemoError::MaxFeeExceeded(Self::MAX_FEE, fee));
+        }
+
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+
+        // Put global index of the previously funded gift code into memo
+        memo_data[..Self::INDEX_DATA_LEN].copy_from_slice(&global_index.to_be_bytes());
+
+        // Put fee into memo
+        memo_data[Self::INDEX_DATA_LEN..(Self::INDEX_DATA_LEN + Self::FEE_DATA_LEN)]
+            .copy_from_slice(&fee.to_be_bytes()[1..]);
+
+        Ok(Self { memo_data })
     }
 
-    /// Get global TxOut index of the cancelled gift code
+    /// Get global index of the TxOut used to fund the gift code
     pub fn cancelled_gift_code_index(&self) -> u64 {
-        u64::from_le_bytes(self.memo_data[0..8].try_into().unwrap())
+        let mut index_bytes = [0u8; 8];
```

Comment:
Optional/Equivalent: Use `INDEX_DATA_LEN`

```suggestion
        let mut index_bytes = [0u8; Self::INDEX_DATA_LEN];
```

Reply:
Done!